core/state/pruner/bloom.go

```diff
@@ -0,0 +1,176 @@
+// potential issue is for fast sync. If we do another fast sync upon the pruned
+// database, it's problematic which will stop the expansion during the syncing.
+// TODO address it @rjl493456442 @holiman @karalabe.
+//
+// After the entire state is generated, the bloom filter should be persisted into
+// the disk. It indicates the whole generation procedure is finished.
+type StateBloom struct {
+	bloom *bloomfilter.Filter
+	done  uint32
+}
+
+// NewStateBloom creates a brand new state bloom for state generation.
+// The optimal bloom filter will be created by the passing "max entries"
+// and the "estimated" maximum collision rate.
+func NewStateBloom(maxElements uint64, probCollide float64) (*StateBloom, error) {
+	bloom, err := bloomfilter.NewOptimal(maxElements, probCollide)
+	if err != nil {
+		return nil, err
+	}
+	log.Info("Initialized state bloom", "size", common.StorageSize(float64(bloom.M()/8)))
+	return &StateBloom{
+		bloom: bloom,
+		done:  0,
+	}, nil
+}
+
+// NewStateBloomWithSize creates a brand new state bloom for state generation.
+// The bloom filter will be created by the passing bloom filter size. According
+// to the https://hur.st/bloomfilter/?n=600000000&p=&m=4096MB&k=3, the parameters
+// are picked so that the false-positive rate for mainnet is low enough.
+func NewStateBloomWithSize(size uint64) (*StateBloom, error) {
+	bloom, err := bloomfilter.New(size*1024*1024*8, 3)
+	if err != nil {
+		return nil, err
+	}
+	log.Info("Initialized state bloom", "size", common.StorageSize(float64(bloom.M()/8)))
+	return &StateBloom{
+		bloom: bloom,
+		done:  0,
+	}, nil
+}
+
+// NewStateBloomFromDisk loads the state bloom from the given file.
+// In this case the assumption is held the bloom filter is complete.
+func NewStateBloomFromDisk(filename string) (*StateBloom, error) {
+	bloom, _, err := bloomfilter.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	return &StateBloom{
+		bloom: bloom,
+		done:  1,
+	}, nil
+}
+
+// Commit flushes the bloom filter content into the disk and marks the bloom
+// as complete.
+func (bloom *StateBloom) Commit(filename string) error {
+	if atomic.CompareAndSwapUint32(&bloom.done, 0, 1) {
+		// Firstly, flush the bloom filter to the temporary file
+		_, err := bloom.bloom.WriteFile(filename + ".tmp")
+		if err != nil {
+			return err
+		}
+		// Secondly, fsync the file. In the bloomfilter library
+		// there is no guarantee the file is actually flushed
+		// into the disk
+		f, err := os.Open(filename + ".tmp")
+		if err != nil {
+			return err
+		}
+		if err := f.Sync(); err != nil {
+			return err
+		}
+		f.Close()
+		// Thirdly, rename the file.
+		if err := rename(filename+".tmp", filename); err != nil {
+			return err
+		}
+		// Lastly, fsync the directory to ensure all pending
+		// rename operations are transferred to disk
+		if err := syncDir(filepath.Dir(filename)); err != nil {
+			return err
+		}
+		return nil
+	}
+	return errors.New("bloom filter is committed")
```

Comment:
Nitpick, instead of doing all the "happy path" nested and have the outer side error out, we should swap them:

```go
if atomic.CompareAndSwapUint32(&bloom.done, 0, 1) {
    return errors.New("bloom filter is committed")
}
[... commit code ...]
```